File: JavaScript/ex/Tree.js
class Tree {
  constructor(value) {
    // constructor로 만든 객체는 트리의 Node가 된다. tree의 자식 노드들을 children으로 할당하여 노드의 자식 노드들을 담을 수 있게 설정한다.
    this.value = value;
    this.children = [];
  }

  // 트리의 삽입 메서드. tree의 자식 노드를 생성 한 후에, 노드의 children에 push
  insertNode(value) {
    const childNode = new Tree(value);
    this.children.push(childNode);
  }

  // 트리 안에 해당 값이 포함되어 있는지 확인하는 메서드. tree에서 value값을 탐색 -> 현재 노드의 value 값이 찾는 값과 일치한다면 return.
  contains(value) {
    if (this.value === value) {
      return true;
    }

    // 노드가 가진 자식 노드를 순회하는 반복문으로 노드의 children 배열을 탐색
    for (let i = 0; i < this.children.length; i++) {
      const childNode = this.children[i];
      if (childNode.contains(value)) {
        return true;
      }
    }
    
    // 전부 탐색했음에도 불구하고 찾지 못했다면 false return.
    return false;
  }
}

File: Python/Tree.py
class Tree:
  def __init__(self, value):
    # 트리의 Node를 생성하며, 자식 노드를 담을 리스트를 초기화한다.
    self.value = value
    self.children = []

  def insert_node(self, value):
    # 새로운 자식 노드를 생성하여 children에 추가한다.
    child_node = Tree(value)
    self.children.append(child_node)

  def contains(self, value):
    # 현재 노드의 값이 찾고자 하는 값과 일치하는지 확인한다.
    if self.value == value:
      return True
    
    # 자식 노드를 순회하며 값이 존재하는지 확인한다.
    for child_node in self.children:
      if child_node.contains(value):
        return True
    
    # 모든 자식 노드를 탐색해도 값이 없으면 False를 반환한다.
    return False
